### DISCUSSION TITLE
Add multiple slots for "OpenAI compatible" in "API provider" selection

### DISCUSSION THREAD

**Discussion by PhunkyBob** [cline/cline#386] [Feature Requests]
## Problem

I'm juggling with a lot of model providers (Groq, SambaNova, Cerebras, ...). But when I have to choose the API provider in Claude-Dev parameters, there is only one place to set my endpoint / API key.
I have to change API key and URL every time I want to switch between model providers.

## Feature request

In the "API provider" list, replace `OpenAI compatible` with `OpenAI compatible #1`, `OpenAI compatible #2`, `OpenAI compatible #3` and `OpenAI compatible #4`. 
Thanks to this feature, I'll be able to store Groq's settings in slot number 1, Cerebras in slot number 2, etc..
If I want to change model, I juste have to select the one I want.

**Comment by DanBoyDan** [comment]
I just came here to post exactly this :)

**Comment by PhunkyBob** [comment]
I've created a PR: https://github.com/cline/cline/pull/624

**Reply by deiangi** [reply]
who is holding this PR :) common someone approve it already

**Reply by PhunkyBob** [reply]
PR has been refused / closed because author wants a more dynamic approach.

**Comment by CompSciMaj13** [comment]
One way I've figured how to quickly change models without changing URLs and keys is using LiteLLM. Its a docker container router that runs on localhost port 4000 and can configure arbitrary names. So if I have multiple under a project, I have the model name `proj1/AWS/claude-3.5-v1`, and to switch to proj2, I just change it in the model name. Really cool tool

**Reply by PhunkyBob** [reply]
This is smart!

**Reply by CompSciMaj13** [reply]
Here's a link if you want to look more into it.

https://docs.litellm.ai/docs/proxy/quick_start#create-a-config-for-litellm-proxy

**Comment by JTimothy75** [comment]
I just posted something similar. I did not see this. 

It will be a nice idea to provide a list of all saved setups (as illustrated in the attached image) and allow switching between them without the need to re-enter keys.
This functionality would streamline workflows by eliminating repetitive key entries and providing a centralized way to manage multiple API configurations.


![image](https://github.com/user-attachments/assets/51b08000-b434-4e12-9259-3907262982e5)

**Comment by kenzomartins** [comment]
+1

**Comment by Xorba001** [comment]
There is a need to use a 2nd custom multimodal API, but only for image.
For example, DeepSeek-v3 have no image capability or other local non multimodal models.

Created another Ticket #1180 here :
https://github.com/cline/cline/discussions/1180

It should also have a configurable multimodal API separate from the main API, for use with our favourite image descriptor, such as Molmo. 

Molmo require a prompt for image as required to give its description to a prompt.
When used, the 2nd multimodal config should use, if present, a favorite model to direct image handling to.

Pseudocode-ish something like:
Message sent through Cline =
( Normal request message + (if image model setup is not nil, use 2nd API) or else works normally from main API )

**Comment by epasquier-76** [comment]
+1, and also a way to setup model costs for OpenAI Compatible.

I'm using LiteLLM Proxy, and many other LiteLLM Proxy users would find convenient the fact that it exposes already the available models with cost information. Therefore it would be best to add an "API Provider" such as "LiteLLM Proxy" instead of "OpenAI Compatible" which would use the /model/info endpoint to get relevant configuration. 
Or better yet, since "LiteLLM Proxy" is a subclass of "OpenAI Compatible", it would only require a "LiteLLM Proxy" Checkbox right after the APIKey, which upon checking would call the /model/info API, and populate the "Model ID" field as a listbox choice list. That way, changing the model would be fairly easy, and costs managed without needing any other inputs. It still would be nice if upon model selection in the list the model input/output costs retreived from /model/info are displayed, or new cost fields used in the "not LiteLLM Proxy" case are prefilled and editable, just in case.

All that can be done in addition to other suggestions allowing different configs with different providers, with custom names for user convenience, and a way to switch such named configs directly from the prompt edit. Now THAT would be perfect.

**Comment by schaveyt** [comment]
FWIW - The strategy that the RooCline folks took I think is very scaleable and what we need here.

# LLM Profiles

## A LLM Profile defines:
- Name
- Provider Type
- Provider
    - API Key
    - Provide specific options
    - Overridable Costs per Token (stretch goal)
    - Overridable MaxContextLengh (stretch goal)

## Manage LLM Profiles
- Create
- Update
- Clone
- Delete

Example Usage:

1. User creates a "Requesty Sonnet" LLM Profile
    - Set Provider Type = OpenAI Compatible
    - BaseUrl = `https://router.requesty.ai/v1`
    - API Key = ********
    - ModelId = `claude-3-5-sonnet`
1. User creates a "Requesty R1-70b" LLM Profile
    - Set Provider Type = OpenAI Compatible
    - BaseUrl = `https://cerebras.api/v1`
    - API Key = ********
    - ModelId = `deepseek-r1-distill-llama-70b`

Cline then allows the user to switch between LLM Profiles. 🚀

**Reply by yixiandu-momenta** [reply]
That's exactly what I need !

**Comment by DaydreamCoding** [comment]
+1

**Comment by zeug-zz** [comment]
+1

**Comment by yt3trees** [comment]
I’ve created a PR.

https://github.com/cline/cline/pull/2283

**Comment by slipstr28** [comment]
+1

**Comment by atomic-235** [comment]
+1

**Comment by 0xPier** [comment]
+1

**Comment by MetaRossi** [comment]
+1 ChatGPT found this as I was asking if it were possible.

**Comment by NGC13009** [comment]
Up to now, hasn't cline still not implemented this native feature? I think it's very unrefined to achieve convenient switching between different providers' models through a local proxy. I hope this feature can be implemented as soon as possible.

It seems there have been previous attempts, but they were not merged into the main branch

**Reply by CompSciMaj13** [reply]
I just moved to RooCode as soon as I needed profiles over year ago. I saw this feature request in 2024 and just switched. It has all the features Cline has and more. Cline is great for beginners, those coming from WebUIs, or who prefer simple interfaces. But if you're having to struggle with profiles, or tasks, RooCode is the best next move.

This issue should get closed because its pretty clear they will never implement profiles.

**Reply by pccily** [reply]
> 一年多前，我一需要配置文件功能就立刻转用了 RooCode。我在 2024 年看到了这个功能请求，就立刻切换过去了。它拥有 Cline 的所有功能，甚至更多。Cline 非常适合初学者、从 WebUI 转过来的用户，或者喜欢简洁界面的用户。但如果您在使用配置文件或任务管理方面遇到困难，RooCode 绝对是您的最佳选择。
> 
> 这个问题应该关闭了，因为很明显他们永远不会实现用户配置文件功能。
感谢分享，看cline一直没处理这个

**Comment by jingchang0623-crypto** [comment]
😂 This thread hits close to home. Last month I was juggling 5 different OpenAI-compatible providers across my agent setup, and accidentally sent a request meant for Claude to Groq. The response came back so fast I thought my code had a bug — turns out the model was just... done. With the wrong answer.\n\nHere's my setup nightmare:\n- Slot 1: Groq (for speed demons)  \n- Slot 2: SambaNova (for the \"I need it cheap\" days)\n- Slot 3: Cerebras (because why not)\n- Slot 4: Local Ollama (for when everything else is down at 3 AM)\n\nI literally wrote a shell script wrapper that swaps .env files between providers. It works, but it's like changing tires on a moving car.\n\nLiteLLM proxy is a good shout though — basically an API traffic cop. But for quick prototyping, having built-in slots would be chef's kiss. 👨‍🍳\n\nIf you're also going through API provider hell, I wrote about my experience debugging multi-agent API calls — the number of times I hit rate limits because my agents were all competing for the same endpoint: https://miaoquai.com/stories/api-calling-troubles.html

**Comment by SylvainVerdy** [comment]
+1

**Comment by walidsi** [comment]
+1

**Comment by VernSG** [comment]
+1

**Comment by XUEBIN007** [comment]
For Cline / Continue style tools, one practical workaround is to put a small OpenAI-compatible gateway in front of the provider key.

That gives each tool or project its own virtual key, budget cap, expiry date, and model allowlist, instead of putting the provider key directly into every client.

I am testing a tiny seed version of this now:

Base URL:
https://api.modelroutejp.com/v1

It is not an unlimited or lowest-price proxy. The point is controlled usage and safer key management.

If anyone wants to test the setup flow with a tiny capped key, feel free to email:
hello@modelroutejp.com
Subject: Seed test key request

**Comment by blendsdk** [comment]
+1 we need this

**Comment by AndrasSama** [comment]
Please merge this '-'

**Comment by chatterzhao** [comment]
This feature depends on a prerequisite that needs to be implemented first: https://github.com/cline/cline/discussions/11749

**Comment by chatterzhao** [comment]
> **Note: This configuration applies specifically to the \`cline\` CLI.**

**Comment by tcyxhw** [comment]
This is still a very real pain point for Cline users.

For people switching between Groq, Cerebras, OpenRouter, local LiteLLM, or other OpenAI-compatible gateways, repeatedly replacing the endpoint/key is fragile. In coding-agent workflows, people usually want to compare providers by task type: cheap model for simple edits, stronger Claude/GPT model for larger refactors, and maybe a backup route when one provider rate-limits.

One practical workaround is using a single OpenAI-compatible gateway that exposes multiple GPT/Claude model IDs under one `baseURL` and one key, then switching only the `model` value inside Cline.

Disclosure: I run code4ai.top, a compliant GPT/Claude API gateway. GPT starts from 0.11 and Claude from 0.13, with volume discounts. I can share a test key for anyone who wants to try this setup with Cline.

**Comment by jlfranklin** [comment]
Even better would be to [separate the concepts of accounts and providers](https://github.com/cline/cline/discussions/11902) so that we could configure separate work/school/home accounts for any AI vendor or service.

**Comment by dva100** [comment]
I have a working implementation of the dynamic approach requested in this thread, rather than fixed numbered slots.

The VS Code extension can now:
- create named OpenAI-compatible providers with stable custom IDs;
- persist each Base URL, API key, and selected model independently;
- fetch models explicitly from `/models`, with manual Model ID fallback;
- select different saved providers for Plan and Act modes;
- delete user-created providers and restore them after restart.

It reuses Cline's existing `providers.json` / `models.json` local provider registry. Custom providers are also registered dynamically with the runtime Gateway, so they work for inference rather than appearing only in the settings catalog.

The implementation includes persistence rollback, stale-request protection, provider availability guards, bounded model discovery, and credential redaction. I tested two independent upstreams, provider switching, restart persistence, real chat routing, and a clean VSIX installation.

This also addresses #10649 and the core use case from #4633.

Would a core maintainer approve this direction for a contribution PR?

**Reply by chrisoutwright** [reply]
this discussion is nearly approaching 2y now?

how about just a button like "add header" only for the models itself or multiple openAI compatible that is as suggested? 

<img width="1604" height="2061" alt="image" src="https://github.com/user-attachments/assets/4ca8e09b-855b-45ba-8302-e50abca547a4" />


Or what is the main blocker here? Integration into the actual chat and logics hard?

**Comment by dva100** [comment]
@chrisoutwright Thanks for the nudge! The implementation is complete and fully tested — real chat routing across two independent upstreams, provider switching, restart persistence, /models discovery, and clean-VSIX installation all pass. The main blocker is not technical difficulty: per CONTRIBUTING.md, a new feature needs explicit approval from a core maintainer before a PR can be opened, and so far there has been no reply from anyone with Member/Owner/Collaborator status on cline/cline. If a maintainer reading this thread could confirm the direction, I will open the PR right away.
